### DISCUSSION TITLE
Update variable after pacman upgrade

### DISCUSSION THREAD

**Discussion by Aiz0** [elkowar/eww#511] [Q&A] [answered]
I'm making a bar and I have an icon that is displayed if I have any packages that can be updated.

minimal example

`
(box :visible {package-updates != 0} "update your packages")
`

I've defined a `defpoll` variable that updates every 2 hours.
`
(defpoll package-updates :interval "2h" :initial 0 "scripts/pacman.fish")
`

This means that unless the variable is updated, the icon will be display for upwards of 2 hours before disappearing. 

With `waybar` I use this pacman hook to solve this problem.
```
# pacman hook to trigger an update of the custom pacman waybar module

[Trigger]
Operation = Upgrade
Type = File
Target = *

[Action]
Description = waybar: refresh pacman module
When = PostTransaction
Exec = /usr/bin/pkill -SIGRTMIN+1 waybar
Depends = waybar
```
This refreshes the custom pacman module which has the `signal` option set to 1

Can I do something similar to that in eww to update the variable?

**Comment by viandoxdev** [comment]
Does that not work ?
```bash
eww update package-updates=0
```

**Reply by Aiz0** [reply]
yes but only as normal user, the commands run by the hook are run as root.

`eww update package-updates=0` fails because `/root/.config/eww does not exist`

setting the configuration directory with -c results in a different error
`ERROR eww::error_handling_ctx > Failed to connect to daemon`

I've also tried using it with `sudo -u` to run as normal user but that also fails to connect to the daemon.

**Reply by Aiz0** [reply]
Running `sudo -u aiz --preserve-env eww update package-updates=0` from shell seems to work.
I assume preserving the environment every time i upgrade my packages isn't a good idea though.

**Reply by viandoxdev** [reply]
Well I think you should be ok with calling eww like so `XDG_RUNTIME_DIR=/run/user/1000 eww -c <path to your config> update <...>` in your pacman hook, this seemed to work on my computer when the daemon was already started by the user. You might want to make sure that XDG_RUNTIME_DIR is right tho.

**Reply by Aiz0** [reply]
That worked, Thanks!

Since I'm always running pacman with either `sudo` or `doas` I can get the runtime dir with
`XDG_RUNTIME_DIR=/run/user/$(id -u $DOAS_USER $SUDO_USER)`
This should ensure `XDG_RUNTIME_DIR` is correct.

To make this better, setting `XDG_CONFIG_HOME` would be nice, though getting that would have to depend on how the users sets it, if they do at all.
For now specifying path to config works.

**Comment by Aiz0** [comment, marked as answer]
I've gotten it to work nicely with this pacman hook.
`runuser` seems to get `XDG_CONFIG_HOME` but not `XDG_RUNTIME_DIR`.

This can probably be simplified or made more robust but it works great for now.

```
# pacman hook to trigger an update of the "package-updates" variable

[Trigger]
Operation = Upgrade
Type = File
Target = *

[Action]
Description = eww: update package-updates variable
When = PostTransaction
Exec = /bin/sh -c "runuser -u $DOAS_USER $SUDO_USER -- sh -c 'XDG_RUNTIME_DIR=/run/user/$(id -u) eww update package-updates=0'"
Depends = eww
```

**Reply by Aiz0** [reply]
I realised that running the script again would be better due there being many situations where an upgrade is made but other packages can still be updated.
```
Exec = /bin/sh -c "runuser -u $DOAS_USER $SUDO_USER -- sh -c 'export XDG_RUNTIME_DIR=/run/user/$(id -u); eww update package-updates=$($XDG_CONFIG_HOME/eww/scripts/pacman.fish)'"
```
